[FIX] Calculate correctly max payload size per UDP packet (#2677)
Haivision/srt: docs/API/API-socket-options.md

| OptName              | Since | Restrict | Type       |  Units  | Default  | Range  | Dir | Entity |
| -------------------- | ----- | -------- | ---------- | ------- | -------- | ------ | --- | ------ |
| `SRTO_MSS`           |       | pre-bind | `int32_t`  | bytes   | 1500     | 116..  | RW  | GSD    |

Maximum Segment Size. This value represents the maximum size of a UDP packet
sent by the system. Therefore the value of `SRTO_MSS` must not exceed the
values of `SRTO_UDP_SNDBUF` or `SRTO_UDP_RCVBUF`. It is used for buffer

> allocation and rate calculation using packet counter assuming fully filled

Reviewer: ```suggestion
allocation and rate calculation using a packet counter that assumes fully filled
```

packets. 

This value is a sum of:

* IP header (20 bytes for IPv4 or 32 bytes for IPv6)
* UDP header (8 bytes)
* SRT header (16 bytes)
* remaining space (as the maximum payload size available for a packet)

> For the default 1500 the "remaining space" part is effectively 1456 for IPv4

Reviewer: ```suggestion
For the default 1500 the "remaining space" is effectively 1456 for IPv4
```

and 1444 for IPv6.

Note that the IP version used here is not the domain of the underlying UDP
socket, but the in-transmission IP version. This is effectively IPv4 in the
following cases:

* when the current socket's binding address is of IPv4 domain
* when the peer's address is an IPv6-mapped-IPv4 address

The IPv6 transmission case is assumed only if the peer's address is a true IPv6 address 
(not IPv4 mapped). It is then not possible to determine the payload size limit 
until the connection is established. SRT operations that must allocate any 
resources according to this value prior to connecting will assume IPv4 transmission 

> because this way, in the worst case, they allocate more space than needed .

Reviewer: ```suggestion
because this way, in the worst case, they allocate more space than needed.
```

is recommended that this value be set to the maximum MTU size of
the network device that you will use for the connection.

The recommendations for the value of `SRTO_MSS` differ between file and live modes.

In live mode a single call to the `srt_send*` function may only send data
that fits in one packet. This size is defined by the `SRTO_PAYLOADSIZE`
option (defult: 1316), and it is also the size of the data in a single UDP
packet. To save memory space, you may want then to set `SRTO_MSS` in live mode to
a value for which the "remaining space" matches the `SRTO_PAYLOADSIZE` value (for
the default value of 1316 this will be 1360 for IPv4 and 1372 for IPv6). For security reasons, 
this is not done by default: it may potentially lead to the inability to read an incoming UDP 
packet if its size is for some reason bigger than the negotiated MSS, which may in turn lead 
to unpredictable behaviour and hard-to-detect errors. You should set such a value only if 
the peer is trusted (that is, you can be certain that you will never receive an oversized UDP 
packet over the link used for the connection). You should also consider the limitations of
`SRTO_PAYLOADSIZE`.

> In the file mode `SRTO_PAYLOADSIZE` has a special value 0 that means no limit
> for one single packet sending, and therefore bigger portions of data are

Reviewer: ```suggestion
In file mode `SRTO_PAYLOADSIZE` has a special value 0 that means no limit
for sending a single packet, and therefore bigger portions of data are
```

internally split into smaller portions, each one using the maximum available
"remaining space". The best value of `SRTO_MSS` for this case is then equal to
the current network device's MTU size. Setting a greater value is possible
(maximum for the system API is 65535), but it may lead to packet fragmentation
on the system level. This is highly unwanted in SRT because:

* SRT also performs its own fragmentation, so it would be counter-productive
* It would use more system resources to no advantage
* SRT is unaware of it, so the resulting statistics would be slightly misleading

System-level packet fragmentation cannot be reliably turned off,

> so safest approach is to avoid it by using appropriate parameters.

Reviewer: ```suggestion
so the safest approach is to avoid it by using appropriate parameters.
```

values of `SRTO_UDP_SNDBUF` or `SRTO_UDP_RCVBUF`. It is used for buffer
allocation and rate calculation using packet counter assuming fully filled
packets. 

This value is a sum of:

* IP header (20 bytes for IPv4 or 32 bytes for IPv6)
* UDP header (8 bytes)
* SRT header (16 bytes)
* remaining space (as the maximum payload size available for a packet)

For the default 1500 the "remaining space" part is effectively 1456 for IPv4
and 1444 for IPv6.

Note that the IP version used here is not the domain of the underlying UDP
socket, but the in-transmission IP version. This is effectively IPv4 in the
following cases:

* when the current socket's binding address is of IPv4 domain
* when the peer's address is an IPv6-mapped-IPv4 address

The IPv6 in-transmission IP version is assumed only if the peer's address
is a true IPv6 address (not IPv4 mapped). It is then not possible to determine
what the payload size limit until the connection is established. Parts of the
SRT facilities that must allocate any resources according to this value prior
to connecting are using the layout as per IPv4 because this way they allocate
more space than needed in the worst case.

> This value can be set on both connection parties independently, but after
> connection this option gets an effectively negotiated value, which is the less
> one from both parties. If this effective value is too small on any of the
> connection peers, the connection is rejected (or late-rejected on the caller
> side).

Reviewer: ```suggestion
This value can be set on both connection parties independently, but after
connection `SRTO_MSS` gets a negotiated value, which is the lesser
of the two. If this effective value is too small for either of the
connection peers, the connection is rejected (or late-rejected on the caller
side).
```

This value then effectively controls:

* The maximum size of the payload in a single UDP packet ("remaining space"). 

* The size of the memory space allocated for a single packet in the sender
and receiver buffers. This value is equal to "SRT header" + "remaining space"
in the IPv4 layout case (1472 bytes per packet for MSS=1500). The reason for it
is that some buffer resources are allocated prior to the connection, so this
value must fit both IPv4 and IPv6 for buffer memory allocation.

> The default value 1500 matches the standard MTU size for network devices. It
> is recommended that this value be set at maximum to the value of MTU size of
> the network device that you will use for connection.

Reviewer: ```suggestion
The default value of 1500 corresponds to the standard MTU size for network devices. It
is recommended that this value be set to the maximum MTU size of
the network device that you will use for the connection.
```

> Detailed recommendations for this value differ in the file and live mode.

Reviewer: ```suggestion
The recommendations for the value of `SRTO_MSS` differ between file and live modes.
```

> In the live mode a single call to `srt_send*` function may only send data
> that fit in one packet. This size is defined by the `SRTO_PAYLOADSIZE`
> option (defult: 1316) and it is also the size of the data in a single UDP
> packet. To save memory space, you may want then to set MSS in live mode to
> a value for which the "remaining space" matches `SRTO_PAYLOADSIZE` value (for
> default 1316 it will be 1360 for IPv4 and 1372 for IPv6). This is not done by
> default for security reasons: this may potentially lead to inability to read an
> incoming UDP packet if its size is by some reason bigger than the negotiated MSS.
> This may lead to misleading situations and hard to detect errors. You should
> set such a value only if the peer is trusted (that is, you can be certain that
> it will never come to a situation of having received an oversized UDP packet
> over the link used for the connection). See also limitations for
> `SRTO_PAYLOADSIZE`.

Reviewer: ```suggestion
In live mode a single call to the `srt_send*` function may only send data
that fits in one packet. This size is defined by the `SRTO_PAYLOADSIZE`
option (defult: 1316), and it is also the size of the data in a single UDP
packet. To save memory space, you may want then to set `SRTO_MSS` in live mode to
a value for which the "remaining space" matches the `SRTO_PAYLOADSIZE` value (for
the default value of 1316 this will be 1360 for IPv4 and 1372 for IPv6). For security reasons, 
this is not done by default: it may potentially lead to the inability to read an incoming UDP 
packet if its size is for some reason bigger than the negotiated MSS, which may in turn lead 
to unpredictable behaviour and hard-to-detect errors. You should set such a value only if 
the peer is trusted (that is, you can be certain that you will never receive an oversized UDP 
packet over the link used for the connection). You should also consider the limitations of
`SRTO_PAYLOADSIZE`.
```

> In the file mode `SRTO_PAYLOADSIZE` has a special value 0 that means no limit
> for one single packet sending, and therefore bigger portions of data are
> internally split into smaller portions, each one using the maximum available
> "remaining space". The best value of `SRTO_MSS` for this case is then equal to
> the current network device's MTU size. Setting a greater value is possible
> (maximum for the system API is 65535), but it may lead to packet fragmentation
> on the system level. This is highly unwanted in SRT because:

Reviewer: ```suggestion
In file mode `SRTO_PAYLOADSIZE` has a special value 0 that means there is no limit
for sending a single packet, and therefore bigger portions of data are
internally split into smaller portions, each one using the maximum available
"remaining space". The best value of `SRTO_MSS` for this case is then equal to
the current network device's MTU size. Setting a greater value is possible
(maximum for the system API is 65535), but it may lead to packet fragmentation
on the system level. This is highly unwanted in SRT because:
```

Author: ```suggestion
In the file mode `SRTO_PAYLOADSIZE` has a special value 0 that means there is
no size limit for a single `srt_send*` call, and bigger portions of data are
internally split into smaller portions, each one using the maximum available
"remaining space" of a packet. The best value of `SRTO_MSS` for this case is then equal to
the current network device's MTU size. Setting a greater value is possible
(maximum for the system API is 65535), but it may lead to packet fragmentation
on the system level. This is highly unwanted in SRT because:
```

Reviewer: ^^ all good.

> * Here SRT does also its own fragmentation, so it would be counter-productive
> * It would use more system resources with no advantage
> * SRT is unaware of it, so the statistics will be slightly falsified

Reviewer: ```suggestion
* SRT also performs its own fragmentation, so it would be counter-productive
* It would use more system resources to no advantage
* SRT is unaware of it, so the resulting statistics would be slightly misleading
```

> The system-level packet fragmentation cannot be however reliably turned off;
> the best approach is then to avoid it by using appropriate parameters.

Reviewer: ```suggestion
System-level packet fragmentation cannot be reliably turned off,
so safest approach is to avoid it by using appropriate parameters.
```

| -------------------- | ----- | -------- | ---------- | ------- | -------- | ------ | --- | ------ |
| `SRTO_PAYLOADSIZE`   | 1.3.0 | pre      | `int32_t`  | bytes   | \*       | 0.. \* | W   | GSD    |

> Sets the data limitation mode and the maximum data size for sending at once.
> 
> The default value is 0 in the file mode and 1316 in live mode (this is one of
> the options modified together with `SRTO_TRANSTYPE`).
> 
> If the value is 0, this means a "file mode", in which the call to `srt_send*`
> is not limited to a size fitting in one single packet, that is, the supplied
> data will be split into multiple pieces fitting in a single UDP packet, if
> necessary, as well as every packet will use the maximum space available
> in a UDP packet (except the last in the stream or in the message) according to
> the `SRTO_MSS` setting and others that may influence this size (such as
> `SRTO_PACKETFILTER` and `SRTO_CRYPTOMODE`).
> 
> If the value is greater than 0, this means a "live mode", and the value
> defines the maximum size of:
> 
> * the single call to a sending function (`srt_send*`)
> * the payload supplied in every single data packet
> 
> This value can be set to a greater value than the default 1316, but the maximum
> possible value is limited by the following factors:
> 
> * 1500 is the default MSS (see `SRTO_MSS`), including headers, which are:
>    * 20 bytes for IPv4 or 32 bytes for IPv6
>    * 8 bytes for UDP
>    * 16 bytes for SRT
> 
> This alone gives the limit of 1456 for IPv4 and 1444 for IPv6. This limit may
> be however further decreased in the following cases:
> 
> * 4 bytes reserved for FEC, if you use the builtin FEC packet filter (see `SRTO_PACKETFILTER`)
> * 16 bytes reserved for the authentication tag, if you use AES GCM (see `SRTO_CRYPTOMODE`)
> 
> **WARNING**: The option setter will reject the setting if this value is too
> great, but note that not every limitation can be checked prior to connection.
> This includes:
> 
> * MSS defined by the peer, which may override MSS set in the agent
> * The in-transmission IP version - see `SRTO_MSS` for details

Reviewer: ```suggestion
Sets the mode that determines the limitations on how data is sent, including the maximum 
size of payload data sent within a single UDP packet.

The default values are 0 (corresponding to file mode) and 1316 (in live mode, which is one 
of the options modified together with `SRTO_TRANSTYPE`).

In file mode (`SRTO_PAYLOAD` = 0) the call to `srt_send*`
is not limited to the size of a single packet. If necessary, the supplied
data will be split into multiple pieces, each fitting into a single UDP packet. Every data payload  
(except the last one in the stream or in the message) will use the maximum space available in 
a UDP packet, as determined by `SRTO_MSS` and other settings that may influence this size 
(such as [`SRTO_PACKETFILTER`](#SRTO_PACKETFILTER) and 
[`SRTO_CRYPTOMODE`](#SRTO_CRYPTOMODE)`SRTO_CRYPTOMODE`).

In live mode  (`SRTO_PAYLOAD` > 0) the value defines the maximum size of:

* a single call to a sending function (`srt_send*`)
* the payload supplied in each data packet

This value can be set to a greater value than the default 1316, but the maximum
possible value is limited by the following factors:

* 1500 bytes is the default MSS (see  [SRTO_MSS](#SRTO_MSS)), including headers, which occupy:
   * 20 bytes for IPv4, or 32 bytes for IPv6
   * 8 bytes for UDP
   * 16 bytes for SRT

This alone gives a limit of 1456 for IPv4 and 1444 for IPv6. This limit may
be further decreased in the following cases:

* 4 bytes reserved for FEC, if you use the built in FEC packet filter (see [`SRTO_PACKETFILTER`](#SRTO_PACKETFILTER))
* 16 bytes reserved for the authentication tag, if you use AES GCM (see [`SRTO_CRYPTOMODE`](#SRTO_CRYPTOMODE))

**WARNING**: The party setting the options will reject a value that is too
large, but note that not every limitation can be checked prior to connection.
This includes:

* the MSS value defined by a peer, which may override the MSS set by an agent
* the in-transmission IP version (see [SRTO_MSS](#SRTO_MSS) for details)
```

Author: ```suggestion
Sets the mode that determines the limitations on how data is sent, including the maximum 
size of payload data sent within a single UDP packet.

The default value is 1316 in live mode (which is default) and 0 in file mode (when file
mode is set through the `SRTO_TRANSTYPE` option).

In file mode (`SRTO_PAYLOADSIZE` = 0) the call to `srt_send*` is not limited to the size
of a single packet. If necessary, the supplied data will be split into multiple pieces,
each fitting into a single UDP packet. Every data payload (except the last one in the
stream or in the message) will use the maximum space available in a UDP packet,
as determined by `SRTO_MSS` and other settings that may influence this size 
(such as [`SRTO_PACKETFILTER`](#SRTO_PACKETFILTER) and 
[`SRTO_CRYPTOMODE`](#SRTO_CRYPTOMODE)).

In live mode (`SRTO_PAYLOADSIZE` > 0) the value defines the maximum size of:

* a single call to a sending function (`srt_send*`)
* the payload supplied in each data packet

as well as the minimum size of the buffer used for the `srt_recv*` call.

This value can be set to a greater value than the default 1316, but the maximum
possible value is limited by the following factors:

* 1500 bytes is the default MSS (see [`SRTO_MSS`](#SRTO_MSS)), including headers, which occupy:
   * 20 bytes for IPv4, or 32 bytes for IPv6
   * 8 bytes for UDP
   * 16 bytes for SRT

This alone gives a limit of 1456 for IPv4 and 1444 for IPv6. This limit may
be further decreased in the following cases:

* 4 bytes reserved for FEC, if you use the built in FEC packet filter (see [`SRTO_PACKETFILTER`](#SRTO_PACKETFILTER))
* 16 bytes reserved for the authentication tag, if you use AES GCM (see [`SRTO_CRYPTOMODE`](#SRTO_CRYPTOMODE))

**WARNING**: The party setting the options will reject a value that is too
large, but note that not every limitation can be checked prior to connection.
This includes:

* the MSS value defined by a peer, which may override the MSS set by an agent
* the in-transmission IP version (see [SRTO_MSS](#SRTO_MSS) for details)
```

Reviewer: ^^ all good.
